src/RadixDlt.NetworkGateway.PostgresIntegration/Queries/KeyValueStoreQueries.cs

```diff
@@ -0,0 +1,430 @@
+        byte networkId,
+        KeyValueStoreKeysQueryConfiguration queryConfiguration,
+        CancellationToken token = default)
+    {
+        var cd = new CommandDefinition(
+            commandText: @"
+WITH vars AS (
+    SELECT
+        @keyValueStoreEntityId AS key_value_store_entity_id,
+        CAST(@useCursor AS bool) AS use_cursor,
+        ROW(CAST(@cursorStateVersion AS bigint), CAST(@cursorDefinitionId  AS bigint)) AS start_cursor_inclusive,
+        @atLedgerState AS at_ledger_state,
+        @definitionReadLimit AS definition_read_limit,
+        @pageLimit as page_limit
+),
+definitions_with_cursor AS (
+     SELECT
+         d.*,
+         (d.from_state_version, d.id) AS cursor
+     FROM key_value_store_entry_definition d
+)
+SELECT
+    -- entity data
+    vars.key_value_store_entity_id AS KeyValueStoreEntityId,
+
+    -- totals
+    COALESCE(entity_totals.total_entries_excluding_deleted, 0) AS TotalEntriesExcludingDeleted,
+    COALESCE(entity_totals.total_entries_including_deleted, 0) AS TotalEntriesIncludingDeleted,
+
+    -- entries
+    CASE WHEN COALESCE(filter_out, TRUE) THEN NULL ELSE entries_per_entity.definition_id END AS DefinitionId,
+    CASE WHEN COALESCE(filter_out, TRUE) THEN NULL ELSE entries_per_entity.key END AS Key,
+    CASE WHEN COALESCE(filter_out, TRUE) THEN NULL ELSE entries_per_entity.key_first_seen_state_version END AS KeyFirstSeenStateVersion,
+
+    CASE WHEN COALESCE(filter_out, TRUE) THEN NULL ELSE entries_per_entity.value END AS Value,
+    CASE WHEN COALESCE(filter_out, TRUE) THEN NULL ELSE entries_per_entity.is_locked END AS IsLocked,
+    CASE WHEN COALESCE(filter_out, TRUE) THEN NULL ELSE entries_per_entity.is_deleted END AS IsDeleted,
+    CASE WHEN COALESCE(filter_out, TRUE) THEN NULL ELSE entries_per_entity.last_updated_state_version END AS LastUpdatedStateVersion,
+
+    -- cursor
+    COALESCE(entries_per_entity.filter_out, TRUE) AS FilterOut,
+    COALESCE(entries_per_entity.is_last_candidate, TRUE) AS IsLastCandidate,
+    next_cursor_inclusive AS NextCursorInclusive
+FROM vars
+
+ -- Totals
+ LEFT JOIN LATERAL (
+    SELECT
+        t.total_entries_excluding_deleted AS total_entries_excluding_deleted,
+        t.total_entries_including_deleted AS total_entries_including_deleted
+    FROM key_value_store_totals_history t
+    WHERE t.entity_id = vars.key_value_store_entity_id AND t.from_state_version <= vars.at_ledger_state
+    ORDER BY t.from_state_version DESC
+    LIMIT 1
+) entity_totals ON TRUE
+
+-- entries_per_entity
+LEFT JOIN LATERAL (
+    SELECT
+        definitions.id as definition_id,
+        definitions.key,
+        definitions.key_first_seen_state_version,
+        definitions.is_last_candidate,
+        definitions.cursor,
+        entries.last_updated_state_version,
+        entries.value,
+        entries.is_locked,
+        entries.is_deleted,
+        CASE WHEN (ROW_NUMBER() OVER (ORDER BY definitions.cursor DESC)) = vars.page_limit OR entries.filter_out
```

Comment:
As per other review, possibly the CASE statement is superfluous? (i.e. we could just return the boolean?)

Reply:
fixed.